File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys
input = lambda: sys.stdin.readline().strip()

n, k = map(int, input().split())
arr = list(map(int, input().split()))
arr.sort()

answer = 0
start = 0
end = 1_000_000_000

while start <= end:
    mid = (start + end) // 2
    temp = 0

    for i in arr:
        if i - mid > 0:
            temp += i - mid

    if temp >= k:
        answer = max(answer, mid)
        start = mid + 1
    else:
        end = mid - 1

print(answer)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.util.*;
import java.io.*;

public class Main {

	static int N, M, arr[];

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		st = new StringTokenizer(br.readLine());
		arr = new int[N];
		
		long start = 0;
		long end = 1000000000;
		
		for (int i = 0; i < N; i++) {
			int n = Integer.parseInt(st.nextToken());
			arr[i] = n;
			end = Math.max(end, n);
		}

		while (start <= end) {
			long mid = (start + end) / 2;

			long len = 0;
			for (int i = 0; i < N; i++) {
				long temp = arr[i] - mid;
				if (temp > 0) len += temp;
			}

			if (len >= M) {
				start = mid + 1;
			} else {
				end = mid - 1;
			}
		}
		System.out.println(end);

	}
}